Add docs for per key inference (#28243)
apache/beam: website/www/site/content/en/documentation/sdks/python-machine-learning.md

If you are unsure if your data is keyed, you can also use `MaybeKeyedModelHandler`.

You can also use a `KeyedModelHandler` to load several different models based on their associated key:

```
from apache_beam.ml.inference.base import KeyedModelHandler
keyed_model_handler = KeyedModelHandler([
  KeyModelMapping(['key1'], PytorchModelHandlerTensor(<config1>)),
  KeyModelMapping(['key2', 'key3'], PytorchModelHandlerTensor(<config2>))
])
with pipeline as p:
   data = p | beam.Create([
      ('key1', torch.tensor([[1,2,3],[4,5,6],...])),
      ('key2', torch.tensor([[1,2,3],[4,5,6],...])),
      ('key3', torch.tensor([[1,2,3],[4,5,6],...])),
   ])
   predictions = data | RunInference(keyed_model_handler)
```

The previous example will load a model using `config1` and use that for inference for all examples associated

> with `key1`, and will load a model using `config2` and use that for all examples associated with `key2` and `key3`.

Author: TODO - this still needs to talk about memory pitfalls/memory management

If you are unsure if your data is keyed, you can also use `MaybeKeyedModelHandler`.

> You can also use a `KeyedModelHandler` to load several different models based on their associated key:

Author: Yes, I thought I had already done that. Updated

      ('key2', torch.tensor([[1,2,3],[4,5,6],...])),
      ('key3', torch.tensor([[1,2,3],[4,5,6],...])),
   ])
   predictions = data | RunInference(keyed_model_handler)
```

The previous example will load a model using `config1` and use that for inference for all examples associated
with `key1`, and will load a model using `config2` and use that for all examples associated with `key2` and `key3`.

There are memory risks associated with loading multiple models at once. By default, `KeyedModelHandler` will not
limit the number of models loaded into memory at once. This means that if not all models fit into memory at once,
your pipeline will likely fail with an Out of Memory exception. To avoid this, you can provide a hint about the
maximum number of models loaded at once.=:

```
mhs = [
  KeyModelMapping(['key1'], PytorchModelHandlerTensor(<config1>)),
  KeyModelMapping(['key2', 'key3'], PytorchModelHandlerTensor(<config2>)),
  KeyModelMapping(['key4'], PytorchModelHandlerTensor(<config3>)),
  KeyModelMapping(['key5', 'key6', 'key7'], PytorchModelHandlerTensor(<config4>)),
]
keyed_model_handler = KeyedModelHandler(mhs, max_models_per_worker_hint=2)
```

> The previous example will load at most 2 models per worker at any given time, and will unload models that aren't

Reviewer: > load at most 2 models per worker 

Users might perceive it as a guarantee, and come to us if/when they see a single OOM error.  If this cannot be guaranteed, we can phrase that the upper  ceiling is enforced as best effort. Or mention that there may be some delay between when the model is unloaded and the memory is freed.

Author: I mentioned that there's some delay, let me know if you think the wording is ok

> The previous example will load at most 2 models per worker at any given time, and will unload models that aren't

Author: I haven't yet, but its a good idea I can follow up with

currently being used as needed. Runners that have multiple workers on a given machine will load at most

> `max_models_per_worker_hint*<num workers>` models onto the machine. Make sure you leave enough space for the models

Reviewer: Since worker is such an overloaded term, users might be confused. In Dataflow context, `worker` often refers to the  VM, as in  `max_num_workers` pipeline option.

Here by 'worker' you refer to SDK worker process. In  https://cloud.google.com/dataflow/docs/guides/troubleshoot-oom#sdk-process-memory we call these  'Apache Beam SDK process`. How about you use that or `SDK worker process` instead of `worker`?

Author: Updated references to `SDK worker process`

If you are unsure if your data is keyed, you can also use `MaybeKeyedModelHandler`.

You can also use a `KeyedModelHandler` to load several different models based on their associated key:

```
from apache_beam.ml.inference.base import KeyedModelHandler
keyed_model_handler = KeyedModelHandler([
  KeyModelMapping(['key1'], PytorchModelHandlerTensor(<config1>)),
  KeyModelMapping(['key2', 'key3'], PytorchModelHandlerTensor(<config2>))
])
with pipeline as p:
   data = p | beam.Create([
      ('key1', torch.tensor([[1,2,3],[4,5,6],...])),
      ('key2', torch.tensor([[1,2,3],[4,5,6],...])),
      ('key3', torch.tensor([[1,2,3],[4,5,6],...])),
   ])
   predictions = data | RunInference(keyed_model_handler)
```

> The previous example will load a model using `config1` and use that for inference for all examples associated

Reviewer 2: ```suggestion
The previous example loads a model by using `config1`. That model is then used for inference for all examples associated
```

> with `key1`, and will load a model using `config2` and use that for all examples associated with `key2` and `key3`.

Reviewer 2: ```suggestion
with `key1`. It then loads a model by using `config2`. That model is used for all examples associated with `key2` and `key3`.
```

> There are memory risks associated with loading multiple models at once. By default, `KeyedModelHandler` will not

Reviewer 2: ```suggestion
Loading multiple models at the same times increases the risk of out of memory (OOM) errors. By default, `KeyedModelHandler` doesn't
```

> limit the number of models loaded into memory at once. This means that if not all models fit into memory at once,

Reviewer 2: ```suggestion
limit the number of models loaded into memory at the same time. If the models don't all fit into memory,
```

> your pipeline will likely fail with an Out of Memory exception. To avoid this, you can provide a hint about the

Reviewer 2: ```suggestion
your pipeline will likely fail with an out of memory error. To avoid this issue, provide a hint about the
```

> maximum number of models loaded at once.

Reviewer 2: ```suggestion
maximum number of models that can be loaded at the same time.
```

```
mhs = [
  KeyModelMapping(['key1'], PytorchModelHandlerTensor(<config1>)),
  KeyModelMapping(['key2', 'key3'], PytorchModelHandlerTensor(<config2>)),
  KeyModelMapping(['key4'], PytorchModelHandlerTensor(<config3>)),
  KeyModelMapping(['key5', 'key6', 'key7'], PytorchModelHandlerTensor(<config4>)),
]
keyed_model_handler = KeyedModelHandler(mhs, max_models_per_worker_hint=2)
```

> The previous example will load at most 2 models per SDK worker process at any given time, and will unload models that aren't

Reviewer 2: ```suggestion
The previous example loads at most two models per SDK worker process at any given time. It unloads models that aren't
```

> currently being used as needed. Runners that have multiple SDK worker processes on a given machine will load at most

Reviewer 2: ```suggestion
currently being used. Runners that have multiple SDK worker processes on a given machine load at most
```

> `max_models_per_worker_hint*<num worker processes>` models onto the machine. Make sure you leave enough space for the models

Reviewer 2: ```suggestion
`max_models_per_worker_hint*<num worker processes>` models onto the machine. Leave enough space for the models
```

> and any additional memory needs from other transforms. There may also be some delay between offloading a model and the

Reviewer 2: ```suggestion
and any additional memory needs from other transforms. Because there might be a delay between when a model is offloaded and when the
```

> memory being released, so it is recommended that you leave some additional buffer.

Reviewer 2: ```suggestion
memory is released, it is recommended that you leave additional buffer.
```

> **Note**: If you have many models but a small `max_models_per_worker_hint`, that can lead to _memory thrashing_ where

Reviewer 2: ```suggestion
**Note**: Having many models but a small `max_models_per_worker_hint` can lead to _memory thrashing_, where
```

a large amount of execution time is wasted swapping models in and out of memory. To reduce the likelihood and impact

> of memory thrashing, consider inserting a

Reviewer 2: ```suggestion
of memory thrashing, if you're using a distributed runner, insert a
```

[GroupByKey](https://beam.apache.org/documentation/transforms/python/aggregation/groupbykey/) transform before your

> inference step if you are using a distributed runner. This will ensure that elements with the same key/model are

Reviewer 2: ```suggestion
inference step. This step reduces thrashing by ensuring that elements with the same key and model are
```

> colocated on the same worker, reducing thrashing.

Reviewer 2: ```suggestion
collocated on the same worker.
```
